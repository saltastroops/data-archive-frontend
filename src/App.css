.dimScreen {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  height: 100%;
  width: 100%;
  margin: 0;
  padding: 0;
  background: #000000;
  opacity: 0.5;
  filter: alpha(opacity=30);
  -moz-opacity: 0.3;
  z-index: 101;
}
.signal {
  border: 15px solid #fff;
  border-radius: 30px;
  height: 130px;
  left: 50%;
  margin: -15px 0 0 -15px;
  opacity: 0;
  position: absolute;
  top: 50%;
  width: 130px;

  animation: pulsate 1s ease-out;
  animation-iteration-count: infinite;
}

@keyframes pulsate {
  0% {
    transform: scale(0.1);
    opacity: 0;
  }
  50% {
    opacity: 2;
  }
  100% {
    transform: scale(3);
    opacity: 0;
  }
}

.search-results.table {
  margin-bottom: 20px;
}

.search-results.cell {
  display: flex;
  align-items: center;
  padding-left: 5px;
  white-space: nowrap;
}

.search-results.cell.header {
  background-color: white;
  border-bottom: 1px solid lightgray;
  cursor: pointer;
}

.search-results.cell.header.cart {
  cursor: text;
}

.search-results.cell span {
  overflow: hidden;
  text-overflow: ellipsis;
}

.search-results.observation-header {
  background-color: #f2fffb;
  border-bottom: 1px solid lightgray;
  border-top: 2px solid lightgray;
}

.search-results.file-row-even {
  background-color: rgb(245, 245, 245);
}

.search-results.file-row-odd {
  background-color: white;
}

.ReactVirtualized__Grid {
  outline: none;
}
