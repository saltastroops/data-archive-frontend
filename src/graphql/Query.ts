import gql from "graphql-tag";

// TODO UPDATE
/**
 * Retrieving the user information.
 */
const USER_QUERY = gql`
  query USER_QUERY {
    user {
      familyName
      givenName
      roles
    }
  }
`;

export { USER_QUERY };
