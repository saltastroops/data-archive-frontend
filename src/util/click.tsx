import { ReactWrapper } from "enzyme";
import * as React from "react";

/**
 * Call the onClick method of the element wrapped in a wrapper with a mouse
 * click event.
 *
 * This function is intended for use in unit tests only. It is required because
 * the standard  Enzyme simulate does not seem to work for routing (see
 * https://github.com/ReactTraining/react-router/issues/4337).
 *
 * Properties
 * ----------
 * wrapper:
 *     Enzyme wrapper, which should wrap exactly one element.
 */
export default function click(
  wrapper: ReactWrapper<any, React.Component, any>
) {
  wrapper.props().onClick(new MouseEvent("click"));
}
