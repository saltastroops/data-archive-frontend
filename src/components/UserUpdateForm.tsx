import { validate } from "isemail";
import * as _ from "lodash";
import * as React from "react";
import { Mutation } from "react-apollo";
import styled from "styled-components";
import { UPDATE_USER_MUTATION } from "../graphql/Mutations";
import { USER_QUERY } from "../graphql/Query";
import InputField from "./basicComponents/InputField";

/**
 * Input for the user update form.
 *
 * Properties:
 * -----------
 * affiliation:
 *     The affiliation of the user, such as a university or an institute.
 * email:
 *     Email address. This will be stored in lower case.
 * familyName:
 *     The family name (surname).
 * id:
 *     The user ID different from the currently logged in user.
 * givenName:
 *     The given name (first name).
 * newPassword:
 *     The new password to replace the old password.
 * confirmNewPassword:
 *     Must match with the new password.
 * password:
 *     The password, which must have at least 7 characters.
 * username:
 *     The username, which must not contain upper case letters.
 */
export interface IUserUpdateFormInput {
  affiliation?: string;
  confirmNewPassword?: string;
  email?: string;
  givenName?: string;
  id?: string;
  familyName?: string;
  newPassword?: string;
  password: string;
  username?: string;
}

/**
 * The state of the update form.
 *
 * Properties:
 * -----------
 * errors:
 *     An object of keys and error messages.
 * userInput:
 *     Values input by the user.
 */
interface IUserUpdateFormState {
  errors: Partial<IUserUpdateFormInput> & { responseError?: string };
  userInput: IUserUpdateFormInput;
}

/**
 * The cache for the update form.
 *
 * The user input and errors are cached.
 */
export interface IUserUpdateFormCache {
  errors?: Partial<IUserUpdateFormInput> & { responseError?: string };
  userInput?: IUserUpdateFormInput;
}

interface IUserUpdateFormProps {
  cache?: IUserUpdateFormCache;
}

const validateUpdateField = (updateInput: IUserUpdateFormInput) => {
  // An object to store errors for all fields.
  const errors: Partial<IUserUpdateFormInput> = {};

  // Check if the password is provided.
  if (!updateInput.password.length) {
    errors.password = "Password must be provided.";
  }

  // Check if the new email address is valid.
  if (
    updateInput.email &&
    !validate(updateInput.email, { minDomainAtoms: 2 })
  ) {
    errors.email = "Email address is invalid";
  }

  // Check if the new username contains upper case characters.
  if (
    updateInput.username &&
    updateInput.username !== updateInput.username.toLowerCase()
  ) {
    errors.username = "Username must be in lowercase";
  }

  // Check if the new password is secure enough.
  if (updateInput.newPassword && updateInput.newPassword.length <= 6) {
    errors.newPassword = "Password should be at least 7 characters long";
  }

  // Check if new passwords match
  if (updateInput.newPassword !== updateInput.confirmNewPassword) {
    errors.confirmNewPassword = "Passwords do not match";
  }

  // return an array consisting of error message if any or empty.
  return errors;
};

const Form = styled.form.attrs({
  className: "column is-4 is-offset-4"
})`
  && {
    padding: 1px;
  }
`;

const Heading = styled.h1.attrs({
  className: "title is-3"
})`
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
`;

const ErrorMessage = styled.p.attrs({
  className: "error tile"
})`
  && {
    text-align: left;
    margin: 3px 0 3px 0;
    padding: 2px 0 2px 0;
    background-color: hsl(348, 100%, 61%);
    color: white;
  }
`;

class UserUpdateForm extends React.Component<
  IUserUpdateFormProps,
  IUserUpdateFormState
> {
  public state: IUserUpdateFormState = {
    errors: {},
    userInput: {
      affiliation: "",
      confirmNewPassword: "",
      email: "",
      familyName: "",
      givenName: "",
      id: "",
      newPassword: "",
      password: "",
      username: ""
    }
  };

  /**
   * Populate the state from cached values.
   */
  componentDidMount() {
    if (this.props.cache) {
      this.setState(() => (this.props.cache as any) || {});
    }
  }

  onHandleSubmit = async (e: React.FormEvent<EventTarget>, updateUser: any) => {
    e.preventDefault();

    // Validate the user input fields
    const errors: object = validateUpdateField(this.state.userInput);
    this.updateState({ errors });

    // Check if there is an error, if there is abort updating.
    if (Object.keys(errors).length > 0) {
      return;
    }

    try {
      // If all fields are validated, update
      const updateVariables = { ...this.state.userInput };
      delete updateVariables.confirmNewPassword;
      await updateUser({ variables: updateVariables });

      // Reset the state when updating succeeded
      this.updateState({
        userInput: {
          affiliation: "",
          confirmNewPassword: "",
          email: "",
          familyName: "",
          givenName: "",
          id: "",
          newPassword: "",
          password: "",
          username: ""
        }
      });

      alert("Successfully updated.");
    } catch (error) {
      // We handle the error with a try ... catch rather than relying on Apollo
      // to pass it on to the render function as we need to store it in the
      // cache
      this.updateState({
        errors: {
          ...this.state.errors,
          responseError: error.message
            .replace("Network error: ", "")
            .replace("GraphQL error: ", "")
        }
      });
    }
  };

  onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.name;
    const value = e.target.value;
    // Update form content with the new user input
    this.updateState({
      userInput: {
        ...this.state.userInput,
        [name]: value
      }
    });
  };

  render() {
    const { errors } = this.state;
    const {
      affiliation,
      confirmNewPassword,
      email,
      familyName,
      givenName,
      id,
      newPassword,
      password,
      username
    } = this.state.userInput;

    return (
      <Mutation
        mutation={UPDATE_USER_MUTATION}
        refetchQueries={[{ query: USER_QUERY }]}
      >
        {(updateUser: any, { loading }: any) => {
          return (
            <Form onSubmit={e => this.onHandleSubmit(e, updateUser)}>
              <fieldset disabled={loading} aria-disabled={loading}>
                <Heading>Update your account</Heading>
                {errors.responseError ? (
                  <ErrorMessage>{errors.responseError}</ErrorMessage>
                ) : null}

                {/* ID */}
                <div className="field">
                  <label className="label">
                    User ID
                    <InputField
                      error={errors.id}
                      name={"id"}
                      onChange={this.onInputChange}
                      value={id}
                    />
                  </label>
                </div>

                {/* Given name */}
                <div className="field">
                  <label className="label">
                    Given name (first name)
                    <InputField
                      error={errors.givenName}
                      name={"givenName"}
                      onChange={this.onInputChange}
                      value={givenName}
                    />
                  </label>
                </div>

                {/* Family name */}
                <div className="field">
                  <label className="label">
                    Family name (surname)
                    <InputField
                      error={errors.familyName}
                      name={"familyName"}
                      onChange={this.onInputChange}
                      value={familyName}
                    />
                  </label>
                </div>

                {/* Email address */}
                <div className="field">
                  <label className="label">
                    Email address
                    <InputField
                      error={errors.email}
                      name={"email"}
                      onChange={this.onInputChange}
                      value={email}
                    />
                  </label>
                </div>

                {/* Username */}
                <div className="field">
                  <label className="label">
                    Username
                    <InputField
                      error={errors.username}
                      name={"username"}
                      onChange={this.onInputChange}
                      value={username}
                    />
                  </label>
                </div>

                {/* Affiliation */}
                <div className="field">
                  <label className="label">
                    Affiliation
                    <InputField
                      error={errors.affiliation}
                      name={"affiliation"}
                      placeholder={"E.g. University of Cape Town"}
                      onChange={this.onInputChange}
                      value={affiliation}
                    />
                  </label>
                </div>

                {/* Password */}
                <div className="field">
                  <label className="label">
                    Password
                    <InputField
                      error={errors.password}
                      name={"password"}
                      onChange={this.onInputChange}
                      type={"password"}
                      value={password}
                    />
                  </label>
                </div>

                {/* New password */}
                <div className="field">
                  <label className="label">
                    New password
                    <InputField
                      error={errors.newPassword}
                      name={"newPassword"}
                      onChange={this.onInputChange}
                      placeholder={"At least 7 characters"}
                      type={"password"}
                      value={newPassword}
                    />
                  </label>
                </div>

                {/* Retype New password */}
                <div className="field">
                  <label className="label">
                    Retype New password
                    <InputField
                      error={errors.confirmNewPassword}
                      name={"confirmNewPassword"}
                      onChange={this.onInputChange}
                      type={"password"}
                      value={confirmNewPassword}
                    />
                  </label>
                </div>

                {/* Submit */}
                <button
                  className="button is-success is-fullwidth is-rounded"
                  data-test="update"
                  type={"submit"}
                >
                  {loading ? "Updating..." : "Update"}
                </button>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }

  /**
   * Update the form state and the cache.
   */
  private updateState = (update: object) => {
    this.setState(
      () => update,
      () => {
        if (this.props.cache) {
          this.props.cache.errors = _.cloneDeep(this.state.errors);
          this.props.cache.userInput = _.cloneDeep(this.state.userInput);
        }
      }
    );
  };
}

export default UserUpdateForm;
