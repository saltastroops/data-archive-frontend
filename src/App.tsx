import * as React from "react";
import { Query } from "react-apollo";
import { Redirect, Route, Switch } from "react-router-dom";
import DataRequestsForm from "./components/dataRequest/DataRequestsForm";
import LoginForm, { ILoginFormCache } from "./components/LoginForm";
import NavigationBar from "./components/NavigationBar";
import RegistrationForm, {
  IRegistrationFormCache
} from "./components/RegistrationForm";
import ISearchFormCache from "./components/searchFormComponents/ISearchFormCache";
import SearchPage from "./components/searchFormComponents/SearchPage";
import UserUpdateForm, {
  IUserUpdateFormCache
} from "./components/UserUpdateForm";
import { USER_QUERY } from "./graphql/Query";

interface IUser {
  familyName: string;
  givenName: string;
  isAdmin: boolean;
}

interface IProtectedRouteProps {
  component: any;
  user: IUser | null | undefined;
  [propName: string]: any;
}

interface ICache {
  loginForm: ILoginFormCache;
  registrationForm: IRegistrationFormCache;
  searchForm: ISearchFormCache;
  userUpdateForm: IUserUpdateFormCache;
}

/**
 * A route which redirects to the login page if the user is not logged in.
 */
function ProtectedRoute({
  component: Component,
  user,
  ...rest
}: IProtectedRouteProps) {
  return (
    <Route
      {...rest}
      render={props =>
        user ? (
          <Component {...props} />
        ) : (
          <Redirect
            to={{
              pathname: "/login",
              state: { from: props.location }
            }}
          />
        )
      }
    />
  );
}

interface IAppState {
  user?: IUser;
  screenDimensions: { innerHeight: number; innerWidth: number };
}

/**
 * The data archive.
 */
class App extends React.Component<{}, IAppState> {
  state = {
    screenDimensions: {
      innerHeight: window.innerHeight,
      innerWidth: window.innerWidth
    },
    user: undefined
  };

  private cache: ICache = {
    loginForm: {},
    registrationForm: {},
    searchForm: {},
    userUpdateForm: {}
  };

  public render() {
    return (
      <Query query={USER_QUERY}>
        {({ data, loading }: any) => {
          if (loading) {
            return <p>Loading...</p>;
          }

          const currentUser =
            data && data.user
              ? {
                  familyName: data.user.familyName,
                  givenName: data.user.givenName,
                  isAdmin: data.user.roles.some(
                    (role: string) => role === "ADMIN"
                  )
                }
              : null;

          return (
            <>
              <NavigationBar user={currentUser} />

              <Switch>
                {/* search page */}
                <Route
                  exact={true}
                  path="/"
                  render={() => (
                    <SearchPage
                      cache={this.cache.searchForm}
                      screenDimensions={this.state.screenDimensions}
                    />
                  )}
                />

                {/* registration page */}
                <Route
                  exact={true}
                  path="/register"
                  render={() => (
                    <RegistrationForm cache={this.cache.registrationForm} />
                  )}
                />

                {/* login page */}
                <Route
                  exact={true}
                  path="/login"
                  render={() => <LoginForm cache={this.cache.loginForm} />}
                />

                {/* account details page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/account"
                  component={() => (
                    <h1 className="title">
                      {currentUser
                        ? `${currentUser.givenName} ${
                            currentUser.familyName
                          } Account`
                        : "No User"}
                    </h1>
                  )}
                />

                {/* data requests page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/data-requests"
                  component={() => <DataRequestsForm />}
                />

                {/* cart page */}
                <Route
                  exact={true}
                  path="/cart"
                  component={() => <h1 className="title">Cart page</h1>}
                />

                {/* admin page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/admin"
                  component={() => <h1 className="title">Admin page</h1>}
                />

                {/* update user page */}
                <ProtectedRoute
                  user={currentUser}
                  exact={true}
                  path="/user-update"
                  component={() => (
                    <UserUpdateForm cache={this.cache.userUpdateForm} />
                  )}
                />

                {/* page not found */}
                <Route
                  component={() => <h1 className="title">Page not found</h1>}
                />
              </Switch>
            </>
          );
        }}
      </Query>
    );
  }
}

export default App;
