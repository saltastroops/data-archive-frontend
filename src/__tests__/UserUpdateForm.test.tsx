import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { MemoryRouter } from "react-router";
import wait from "waait";
import App from "../App";
import UserUpdateForm from "../components/UserUpdateForm";
import { UPDATE_USER_MUTATION } from "../graphql/Mutations";
import { USER_QUERY } from "../graphql/Query";
import click from "../util/click";

window.matchMedia = jest.fn().mockImplementation(query => {
  return {
    addListener: jest.fn(),
    matches: false,
    media: query,
    onchange: null,
    removeListener: jest.fn()
  };
});

const alert = jest.fn();
window.alert = alert;

// Helper function for simulating input field value change.
function inputTyping(wrapper: any, name: string, value: string) {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
}

describe("UserUpdateForm Component", () => {
  it("renders the UserUpdateForm having unpopulated props with no errors", () => {
    // UserUpdateForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Expect the snapshot to match the UserUpdateForm component.
    expect(toJson(wrapper.find("form"))).toMatchSnapshot();
  });

  it("displays no errors if submitted inputs are all valid", async () => {
    const mocks = [
      {
        request: {
          query: UPDATE_USER_MUTATION,
          variables: {
            affiliation: "",
            confirmNewPassword: "",
            email: "",
            familyName: "Doe",
            givenName: "John",
            id: "",
            newPassword: "",
            password: "",
            roles: [],
            username: ""
          }
        },
        result: {
          data: {
            user: {}
          }
        }
      },
      {
        request: {
          query: USER_QUERY
        },
        result: {
          data: {
            user: {
              __typename: "User",
              affiliation: "SAAO",
              email: "doe@example.com",
              familyName: "Doe",
              givenName: "John",
              id: "1",
              roles: [],
              username: "john.doe"
            }
          }
        }
      }
    ];

    // UserUpdateForm component wrapper.
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UserUpdateForm />
      </MockedProvider>
    );

    // UserUpdateForm component instance
    const instance: any = wrapper.find("UserUpdateForm").instance();

    // Simulate state change when the affiliation input field value changes.
    inputTyping(wrapper, "affiliation", "Update affiliation");
    // Simulate state change when the confirmNewPassword input field value changes.
    inputTyping(wrapper, "confirmNewPassword", "validupdatepassword");
    // Simulate state change when the emailAddress input field value changes.
    inputTyping(wrapper, "email", "valid@update.address");
    // Simulate state change when the familyName input field value changes.
    inputTyping(wrapper, "familyName", "Update family name");
    // Simulate state change when the givenName input field value changes.
    inputTyping(wrapper, "givenName", "Update given name");
    // Simulate state change when the new password input field value changes.
    inputTyping(wrapper, "newPassword", "validupdatepassword");
    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");
    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "updateusername");

    // Expect the property affiliation of the state to have been updated with the correct value.
    expect(instance.state.userInput.affiliation).toBe("Update affiliation");

    // Expect the property confirmNewPassword of the state to have been updated with the correct value.
    expect(instance.state.userInput.confirmNewPassword).toBe(
      "validupdatepassword"
    );

    // Expect the property emailAddress of the state to have been updated with the correct value.
    expect(instance.state.userInput.email).toBe("valid@update.address");

    // Expect the property familyName of the state to have been updated with the correct value.
    expect(instance.state.userInput.familyName).toBe("Update family name");

    // Expect the property givenName of the state to have been updated with the correct value.
    expect(instance.state.userInput.givenName).toBe("Update given name");

    // Expect the property newPassword of the state to have been updated with the correct value.
    expect(instance.state.userInput.newPassword).toBe("validupdatepassword");

    // Expect the property password of the state to have been updated with the correct value.
    expect(instance.state.userInput.password).toBe("oldpassword");

    // Expect the property username of the state to have been updated with the correct value.
    expect(instance.state.userInput.username).toBe("updateusername");

    // Expect the button not to indicate loading
    expect(wrapper.find('[data-test="update"]').text()).toContain("Update");

    // Simulate the submitting of the form
    wrapper.find('[data-test="update"]').simulate("submit");

    // Expect the button to indicate loading
    expect(wrapper.find('[data-test="update"]').text()).toContain(
      "Updating..."
    );

    // Expect no error message.
    expect(wrapper.find("p").length).toBe(0);
  });

  it("displays an error message if an invalid email address is submitted", async () => {
    // UserUpdateForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Simulate state change when the emailAddress input field value changes
    inputTyping(wrapper, "email", "invalidupdate.email&address");

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Simulate the submitting of the form
    wrapper.find('[data-test="update"]').simulate("submit");

    // Expect an error message
    expect(wrapper.find("p").length).toBe(1);

    // Expect a meaningful error message
    expect(wrapper.find("p").text()).toContain("Email address");
    expect(wrapper.find("p").text()).toContain("invalid");
  });

  it("displays an error message if an invalid username is submitted", () => {
    // UserUpdateForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Simulate state change when the username input field value changes.
    inputTyping(wrapper, "username", "Updateusername");

    // Simulate the submitting of the form.
    wrapper.find('[data-test="update"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(wrapper.find("p").text()).toContain("Username");
    expect(wrapper.find("p").text()).toContain("lowercase");
  });

  it("displays an error message if an invalid password is submitted", () => {
    // UserUpdateForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Simulate state change when the new password input field value changes
    inputTyping(wrapper, "newPassword", "update");

    // Simulate state change when the password input field value changes
    inputTyping(wrapper, "password", "oldpassword");

    // Simulate the submitting of the form
    wrapper.find('[data-test="update"]').simulate("submit");

    // Expect an error message
    expect(wrapper.find("p").length).toBe(2);

    // Expect meaningful error message
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("7 characters");
  });

  it("displays an error message if an invalid confirmed password is submitted", () => {
    // UserUpdateForm component wrapper.
    const wrapper = mount(
      <MockedProvider>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Simulate state change when the new password input field value changes
    inputTyping(wrapper, "newPassword", "validupdatepassword");

    // Simulate state change when the confirmPassword input field value changes
    inputTyping(
      wrapper,
      "confirmNewPassword",
      "validupdatepasswordnotmatching"
    );

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Simulate the submitting of the form.
    wrapper.find('[data-test="update"]').simulate("submit");

    // Expect an error message.
    expect(wrapper.find("p").length).toBe(1);

    // Expect meaningful error message
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("Password");
    expect(
      wrapper
        .find("p")
        .first()
        .text()
    ).toContain("do not");
  });

  it("should cache values and errors", async () => {
    // As we are about to mount the App (rather than just the user update form)
    // we need to provide a user query.
    const mocks = [
      {
        request: {
          query: USER_QUERY
        },
        result: {
          data: {
            user: {
              __typename: "User",
              affiliation: "SAAO",
              email: "john.doe@example.com",
              familyName: "Doe",
              givenName: "John",
              id: "1",
              roles: [],
              username: "john"
            }
          }
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <MemoryRouter>
          <App />
        </MemoryRouter>
      </MockedProvider>
    );

    // Wait for the user query to finish
    await wait(0);
    wrapper.update();

    // Navigate to the user update form
    const userUpdateFormLink = wrapper.find('a[href="/user-update"]').first();
    click(userUpdateFormLink);

    await wait(0);
    wrapper.update();

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Fill in an invalid email
    inputTyping(wrapper, "email", "invalid email");

    // Submit the form
    const submitButton = wrapper.find('[data-test="update"]');
    submitButton.simulate("submit");

    await wait(0);
    wrapper.update();

    // The value has been stored in the state, and there is an error
    const userUpdateFormState: any = wrapper.find("UserUpdateForm").state();
    const emailValue = userUpdateFormState.userInput.email;
    expect(emailValue).toEqual("invalid email");
    const emailErrorMessage = userUpdateFormState.errors.email;
    expect(emailErrorMessage.length).toBeGreaterThan(0);

    // Navigate away from the user update form
    const cartLink = wrapper.find('a[href="/cart"]').first();
    click(cartLink);

    await wait(0);
    wrapper.update();

    // No user update form any longer
    expect(wrapper.find("UserUpdateForm").length).toBe(0);

    // Navigate back to the user update form
    click(userUpdateFormLink);

    await wait(0);
    wrapper.update();

    // The values and errors have been re-inserted
    const newUserUpdateFormState: any = wrapper.find("UserUpdateForm").state();
    expect(newUserUpdateFormState.userInput.email).toEqual(emailValue);
    expect(newUserUpdateFormState.errors.email).toEqual(emailErrorMessage);
  });

  it("successfully calls the user update mutation with the correct content", async () => {
    const updateUser = jest.fn();
    const mocks = [
      {
        request: {
          query: UPDATE_USER_MUTATION,
          variables: {
            affiliation: "",
            email: "",
            familyName: "Doe",
            givenName: "John",
            id: "",
            newPassword: "",
            password: "oldpassword",
            username: ""
          }
        },
        result: () => {
          updateUser();
          return {
            data: { id: "1" } // The return value is not used
          };
        }
      },
      {
        request: {
          query: USER_QUERY
        },
        result: {
          data: {
            user: {
              __typename: "User",
              affiliation: "SAAO",
              email: "john.doe@example.com",
              familyName: "Doe",
              givenName: "John",
              id: "1",
              roles: [],
              username: "john"
            }
          }
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Fill in a given name
    inputTyping(wrapper, "givenName", "John");

    // Fill in a family name
    inputTyping(wrapper, "familyName", "Doe");

    // The mutation has not been called (yet)
    expect(updateUser).not.toHaveBeenCalled();

    // Submit the form
    alert.mockReset();
    const submitButton = wrapper.find('[data-test="update"]');
    submitButton.simulate("submit");

    // The mutation has been called
    await wait(0);
    wrapper.update();
    expect(updateUser).toHaveBeenCalled();

    // An alert with the success message was shown
    expect(alert).toHaveBeenCalledWith("Successfully updated.");
  });

  it("should show an error if submission fails", async () => {
    const mocks = [
      {
        error: new Error("The server is having a coffee break!"),
        request: {
          query: UPDATE_USER_MUTATION,
          variables: {
            affiliation: "",
            email: "",
            familyName: "",
            givenName: "John",
            id: "",
            newPassword: "",
            password: "oldpassword",
            username: ""
          }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <UserUpdateForm />
      </MockedProvider>
    );

    // Simulate state change when the password input field value changes.
    inputTyping(wrapper, "password", "oldpassword");

    // Fill in a given name
    inputTyping(wrapper, "givenName", "John");

    // There is no error (yet)
    expect(wrapper.find("p.error").length).toBe(0);

    // Submit the form
    alert.mockReset();
    const submitButton = wrapper.find('[data-test="update"]');
    submitButton.simulate("submit");

    // There is an error now.
    await wait(50); // 50 ms rather than 0 ms to give the state update a chance to finish
    wrapper.update();
    expect(wrapper.find("p.error").length).toBe(1);
  });
});
