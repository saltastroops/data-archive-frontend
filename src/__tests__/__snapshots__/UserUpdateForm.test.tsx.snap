// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserUpdateForm Component renders the UserUpdateForm having unpopulated props with no errors 1`] = `
.c0.c0 {
  padding: 1px;
}

.c1.c1 {
  text-align: center;
  margin: 20px 0 20px 0;
}

<form
  className="column is-4 is-offset-4 c0"
  onSubmit={[Function]}
>
  <fieldset
    aria-disabled={false}
    disabled={false}
  >
    <styled.h1>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [
              Object {
                "className": "title is-3",
              },
            ],
            "componentStyle": ComponentStyle {
              "componentId": "sc-cvbbAY",
              "isStatic": true,
              "lastClassName": "c1",
              "rules": Array [
                "
  && {
    text-align: center;
    margin: 20px 0 20px 0;
  }
",
              ],
            },
            "displayName": "styled.h1",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-cvbbAY",
            "target": "h1",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <h1
          className="title is-3 c1"
        >
          Update your account
        </h1>
      </StyledComponent>
    </styled.h1>
    <div
      className="field"
    >
      <label
        className="label"
      >
        User ID
        <InputField
          name="id"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="id"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Given name (first name)
        <InputField
          name="givenName"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="givenName"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Family name (surname)
        <InputField
          name="familyName"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="familyName"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Email address
        <InputField
          name="email"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="email"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Username
        <InputField
          name="username"
          onChange={[Function]}
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="username"
              onChange={[Function]}
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Affiliation
        <InputField
          name="affiliation"
          onChange={[Function]}
          placeholder="E.g. University of Cape Town"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="affiliation"
              onChange={[Function]}
              placeholder="E.g. University of Cape Town"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Password
        <InputField
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="password"
              onChange={[Function]}
              type="password"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        New password
        <InputField
          name="newPassword"
          onChange={[Function]}
          placeholder="At least 7 characters"
          type="password"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="newPassword"
              onChange={[Function]}
              placeholder="At least 7 characters"
              type="password"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <div
      className="field"
    >
      <label
        className="label"
      >
        Retype New password
        <InputField
          name="confirmNewPassword"
          onChange={[Function]}
          type="password"
          value=""
        >
          <div
            className="control"
          >
            <input
              className="input "
              name="confirmNewPassword"
              onChange={[Function]}
              type="password"
              value=""
            />
          </div>
        </InputField>
      </label>
    </div>
    <button
      className="button is-success is-fullwidth is-rounded"
      data-test="update"
      type="submit"
    >
      Update
    </button>
  </fieldset>
</form>
`;
